src/ioctls/kvm_vec.rs

```diff
@@ -0,0 +1,394 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use kvm_bindings::__IncompleteArrayField;
+use std::marker::PhantomData;
+use std::mem;
+use std::mem::size_of;
+
+/// Errors associated with the KvmVec struct.
+#[derive(Debug, Clone)]
+pub enum Error {
+    /// The max size has been exceeded
+    SizeLimitExceeded,
+}
+
+// Trait for accessing some properties of certain KVM structures that resemble an array.
+//
+// The kvm API has many structs that resemble the following `T` structure:
+//
+// ```
+// #[repr(C)]
+// struct T {
+//    len: u32,
+//    some_data: u32,
+//    entries: __IncompleteArrayField<F>,
+// }
+// ```
+#[allow(clippy::len_without_is_empty)]
+pub trait KvmArray<F> {
+    /// Get the array length
+    ///
+    fn len(&self) -> u32;
+
+    /// Get the array length as mut
+    ///
+    fn len_mut(&mut self) -> &mut u32;
+
+    /// Get max array length
+    ///
+    fn max_len() -> u32;
+
+    /// Get the array entries
+    ///
+    fn entries(&self) -> &__IncompleteArrayField<F>;
+
+    /// Get the array entries as mut
+    ///
+    fn entries_mut(&mut self) -> &mut __IncompleteArrayField<F>;
+}
+
+/// An adapter that helps in treating a KvmArray more like an actual `Vec`.
+///
+pub struct KvmVec<T: Default + KvmArray<F>, F> {
```

Comment:
You can get rid of type parameter `F` by using the `type` option I described above when defining `KvmArray`.

This would also get rid of the `PhantomData<F>`.

In the couple of places where `F` is used in `impl KvmVec { }`, you can swap it with `<T as KvmEntriesSize>::Entry` (where `Entry` is whatever type name you define in `KvmArray`).

I believe this will make using this structure easier/more intuitive.